Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -1540,17 +1540,46 @@ extension BrowserViewController: URLBarDelegate {
     }
 
     func urlBar(urlBar: URLBarView, didSubmitText text: String) {
-        // If we can't make a valid URL, do a search query.
-        // If we still don't have a valid URL, something is broken. Give up.
-        let engine = profile.searchEngines.defaultEngine
-        guard let url = URIFixup.getURL(text) ??
-                        engine.searchURLForQuery(text) else {
-            log.error("Error handling URL entry: \"\(text)\".")
+        let trimmedText = text.stringByTrimmingCharactersInSet(NSCharacterSet.whitespaceCharacterSet())
+        
+        guard trimmedText.contains(" ") else {
+            submitText(text)
             return
         }
 
-        Telemetry.recordEvent(SearchTelemetry.makeEvent(engine: engine, source: .URLBar))
+        let possibleKeywordQuerySeparatorSpace = trimmedText.characters.indexOf(" ")!
+        let possibleKeyword = trimmedText.substringToIndex(possibleKeywordQuerySeparatorSpace)
+        let possibleQuery = trimmedText.substringFromIndex(possibleKeywordQuerySeparatorSpace.successor())
+        
+        profile.bookmarks.getURLForKeywordSearch(possibleKeyword).uponQueue(dispatch_get_main_queue()) {
+                if let urlString = $0.successValue {
+                    let engine = self.profile.searchEngines.defaultEngine
 
+                    guard let url = engine.searchURLForQuery(possibleQuery, searchTemplate: urlString) else {
```

Comment:
@thebnich : Yes, at this point we need to replace `%s` in URL with the query. However, the problem is when the query contains whitespace. This should get replaced as `%20`.

I used the `OpenSearchEngine.searchURLForQuery()` function because it was already doing something similar (like percent encoding of unallowed characters). If there are characters (not just whitespace) that need to be percent encoded, either in the URL or in the query, they'll be handled here.